iree/turbine/kernel/wave/codegen/read_write.py

```diff
@@ -342,30 +366,37 @@ def _linearize_memref(mem: Value, offsets: tuple[Value | int]) -> Value:
         layout=Attribute.parse("strided<[1], offset: ?>"),
         memory_space=memory_space,
     )
-    return memref_d.reinterpret_cast(
-        resut_type,
-        base,
-        offsets=[offset],
-        sizes=[size_full],
-        strides=[],
-        static_offsets=[dyn_val],
-        static_sizes=[dyn_val],
-        static_strides=[1],
+    return (
+        memref_d.reinterpret_cast(
+            resut_type,
+            mem,
+            offsets=[offset],
+            sizes=[size_full],
+            strides=[],
+            static_offsets=[dyn_val],
+            static_sizes=[dyn_val],
+            static_strides=[1],
+        ),
+        offset_th,
     )
 
 
 def _create_vec_read(
     emitter: WaveEmitter,
+    symbolic_shape: tuple[IndexExpr, ...],
     mem: Value,
     vector_type: IrType,
     start_indices: tuple[Value],
+    start_indices_wg: tuple[Value],
+    start_indices_th: tuple[Value],
     elements_per_thread: int,
     mask: Optional[Value],
     offsets_vec: Optional[Value],
 ) -> Value:
     if mask is None and offsets_vec is None:
         return vector_d.load(vector_type, mem, start_indices)
 
+    is_gather = offsets_vec is not None

```

Comment:
Don't we also need to check the address space since we can have gathers from shared memory? In this case, we cannot use buffer ops right? (Also same for scatters from shared memory)

Reply:
done